File: py/minsk/analysis/syntax/expressions/unary.py
from collections.abc import Iterable
from dataclasses import dataclass

from minsk.analysis.syntax.expression import ExpressionSyntax
from minsk.analysis.syntax.kind import SyntaxKind
from minsk.analysis.syntax.node import SyntaxNode
from minsk.analysis.syntax.token import SyntaxToken


@dataclass(frozen=True)
class UnaryExpressionSyntax(ExpressionSyntax):
    operator_token: SyntaxToken
    operand: ExpressionSyntax

    @property
    def kind(self) -> SyntaxKind:
        return SyntaxKind.UnaryExpression

    @property
    def children(self) -> Iterable[SyntaxNode]:
        yield self.operator_token
        yield self.operand

File: cpp/minsk/source/analysis/syntax/nodes/expressions/unary.cpp
#include "minsk/analysis/syntax/nodes/expressions/unary.hpp"
#include <utility>

using minsk::analysis::syntax::expression_syntax;
using minsk::analysis::syntax::syntax_kind;
using minsk::analysis::syntax::syntax_node;
using minsk::analysis::syntax::syntax_token;
using minsk::analysis::syntax::unary_expression_syntax;

unary_expression_syntax::unary_expression_syntax(
    syntax_token &&operator_token, std::unique_ptr<expression_syntax> operand)
    : m_operator_token(std::move(operator_token)),
      m_operand(std::move(operand)) {}

syntax_kind unary_expression_syntax::kind() const {
  return syntax_kind::unary_expression;
}

std::vector<const syntax_node *> unary_expression_syntax::children() const {
  return {
      &m_operator_token,
      m_operand.get(),
  };
}

const syntax_token &unary_expression_syntax::operator_token() const {
  return m_operator_token;
}

const expression_syntax *unary_expression_syntax::operand() const {
  return m_operand.get();
}
